Android AppSync Extension Library Documentation Improvements (#8217)
aws-amplify/docs: src/pages/[platform]/build-a-backend/data/aws-appsync-apollo-extensions/index.mdx

  };
}

> AWS AppSync Apollo Extensions provide a seamless way to connect to your AWS AppSync backend using Apollo client, an open-source GraphQL client. The included AWS AppSync authorizers make it simple to apply the correct authorization payloads to your GraphQL operations within the Apollo client. The library allows you to connect directly to your AWS AppSync backend, including AppSync backends managed through Amplify Data.

Reviewer: This paragraph feels repetitive - I preferred the old intro with the features listed separately.

Author: I can split back up. I think it was a mistake but "Features" was at an H2 Level where almost the rest of the documentation was contained under Features. This is primarily why I had removed it in the first place.

</InlineFilter>

## Install the AWS AppSync Apollo Extensions library

<InlineFilter filters={["android"]}>

<BlockSwitcher>

> <Block name=" Amplify Data">

Reviewer: We need something to explain what the difference between these tabs are. The comments in the first code block explain it but those are easy to miss.

Author: @mattcreaser Are you suggesting explaining this once in the intro, or modifying the names of the tabs?

Author: Actually, if we go with "With Amplify and "Without Amplify" that clears it up immediately.

Add the `apollo-appsync-amplify` dependency to your app/build.gradle.kts file.

```kotlin title="app/build.gradle.kts"
dependencies {
    // highlight-start
    // Connect Apollo to AppSync, delegating some implementation details to Amplify
    implementation("com.amplifyframework:apollo-appsync-amplify:1.0.0")    
    // highlight-end
}
```

</Block>

> <Block name="AppSync">

Reviewer: AppSync should be "AWS AppSync" throughout.

But how about making these tabs "With Amplify" and "Without Amplify"?

Author: Yeah I like that better.

</InlineFilter>

## Connecting to AWS AppSync with Apollo client

<InlineFilter filters={["android"]}>

<BlockSwitcher>

<Block name=" Amplify Data">
Before you begin, you will need an Amplify Data backend deploy. To get started, see [Set up Data](/[platform]/build-a-backend/data/set-up-data/). 

Once you have deployed your backend and created the `amplify_outputs.json` file, you can use Amplify library to read and retrieve your configuration values with the following steps:

```kotlin
// Use apiKey auth mode, reading configuration from AmplifyOutputs
val connector = ApolloAmplifyConnector(context, AmplifyOutputs(R.raw.amplify_outputs))
val apolloClient = ApolloClient.Builder()
    .appSync(connector.endpoint, connector.apiKeyAuthorizer())
    .build()
```

### AppSync Authorizers

> Depending on the authorization strategy defined on your schema, you can use the corresponding Authorizer. To read more about the strategies and their corresponding auth modes, see [Available authorization strategies](/[platform]/build-a-backend/data/customize-authz/#available-authorization-strategies).

Reviewer: How about this:

> The AWS AppSync Apollo Extensions library provides a number of Authorizer classes to match the various authorization strategies that may be in use in your schema. You should choose the appropriate Authorizer type for your authorization strategy. To read more about the strategies and their corresponding auth modes, see [Available authorization strategies](/[platform]/build-a-backend/data/customize-authz/#available-authorization-strategies).

Some common ones are

*  `publicAPIkey` strategy, `apiKey` authMode, **APIKeyAuthorizer**
*  `guest` strategy, `identityPool` authMode, **IAMAuthorizer**
*  `owner` strategy, `userPool` authMode, **AuthTokenAuthorizer**

If you define multiple authorization strategies within your schema, you will have to create separate Apollo client instances for each Authorizer that you want to use in your app.

#### API_KEY

> An `ApiKeyAuthorizer` can be read from `amplify_outputs.json`, provided with a hardcoded API key or by fetching the key from some source:

Reviewer: ```suggestion
An `ApiKeyAuthorizer` can read the API key from `amplify_outputs.json`, provide a hardcoded API key, or fetch the API key from some source:
```

//highlight-end
```

</Block>

<Block name="AppSync">

> 

Reviewer: Feels like we're missing a piece of text here. There should be something introducing the first code block.

//highlight-end
```

</Block>

> <Block name="Without Amplify">

Reviewer 2: I hesitate with putting prose like this inside the block switcher for a few reasons:
1. it makes linking to specific content difficult
2. nesting headings messes with the document structure -- do they render okay in the TOC?
3. since there are multiple instances of the blockswitcher and they do not sync on the page it can make reading this doc "without amplify" a bit tedious

We can modify the component to fix 3, but I have concerns with 2 if the heading structure is different between the two blocks

1. Navigate to your API on the [AWS AppSync console](https://console.aws.amazon.com/appsync/home)
2. On the left side, select Schema
3. Select the "Export schema" dropdown and download the `schema.json` file.
4. Add this file to your project as directed by [Apollo documentation](https://www.apollographql.com/docs/kotlin/advanced/plugin-recipes#specifying-the-schema-location) 

</InlineFilter>

## Generating Queries, Mutations, and Subscriptions for Apollo client

<InlineFilter filters={["android"]}>

<BlockSwitcher>

<Block name="With Amplify">
### Amplify provided .graphql files
1. Within your Amplify Gen2 backend, run: `npx ampx generate graphql-client-code --format graphql-codegen --statement-target graphql --out graphql`
2. Copy the generated files (`mutations.graphql`, `queries.graphql`, `subscriptions.graphql`) to your `{app}/src/main/graphql` folder as shown in the [Apollo documentation](https://www.apollographql.com/docs/kotlin#getting-started)

### Manual
1. Navigate to the **Queries** tab in your API on the [AWS AppSync console](https://console.aws.amazon.com/appsync/home). Here, you can test queries, mutations, and subscriptions in the GraphQL playground.

> 2. Enter your GraphQL operation (query, mutation, or subscription) in the editor and click **Run** to execute it.

Reviewer 2: ```suggestion
2. Enter your GraphQL operation (query, mutation, or subscription) in the editor and select **Run** to execute it.
```

nit "click" -> "select"
https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md#15-use-carets-to-help-users-navigate-menus-use-select-instead-of-click

2. On the left side, select Schema
3. Select the "Export schema" dropdown and download the `schema.json` file.
4. Add this file to your project as directed by [Apollo Code Generation documentation](https://www.apollographql.com/docs/ios/code-generation/introduction). 

You can alternatively download the introspection schema using the [`fetch-schema`](https://www.apollographql.com/docs/ios/code-generation/codegen-cli#fetch-schema) command with the `amplify-ios-cli` tool.

</InlineFilter>

<InlineFilter filters={["android"]}>

1. Navigate to your API on the [AWS AppSync console](https://console.aws.amazon.com/appsync/home)
2. On the left side, select Schema
3. Select the "Export schema" dropdown and download the `schema.json` file.
4. Add this file to your project as directed by [Apollo documentation](https://www.apollographql.com/docs/kotlin/advanced/plugin-recipes#specifying-the-schema-location) 

</InlineFilter>

## Generating Queries, Mutations, and Subscriptions for Apollo client

<InlineFilter filters={["android"]}>

<BlockSwitcher>

<Block name="With Amplify">
**Amplify provided .graphql files**

> 1. Within your Amplify Gen2 backend, run: `npx ampx generate graphql-client-code --format graphql-codegen --statement-target graphql --out graphql`

Reviewer 2: ```suggestion
1. Within your Amplify Gen 2 backend, run: `npx ampx generate graphql-client-code --format graphql-codegen --statement-target graphql --out graphql`
```

nit, Gen2 -> Gen 2

</InlineFilter>

<InlineFilter filters={["android"]}>

1. Navigate to your API on the [AWS AppSync console](https://console.aws.amazon.com/appsync/home)
2. On the left side, select Schema

> 3. When viewing your schema, there should a “Export schema” drop down. Select this and download the `schema.json` file.

Reviewer: ```suggestion
3. Select the "Export schema" dropdown and download the `schema.json` file.
```

Reviewer: Same on AppSync tab
